File: week_25/06_timeout.go
package main

import (
	"context"
	"fmt"
	"time"
)

// 06. Timeout Pattern in Go

func main() {
	fmt.Println("=== Go Timeout with Context ===")
	fmt.Println()

	// 1. Basic timeout with context
	fmt.Println("1. Basic timeout:")
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	done := make(chan string)
	go func() {
		fmt.Println("  Starting long operation...")
		time.Sleep(1 * time.Second)
		done <- "Completed"
	}()

	select {
	case result := <-done:
		fmt.Printf("  ✓ %s in time\n", result)
	case <-ctx.Done():
		fmt.Println("  ✗ Operation timed out!")
	}
	fmt.Println()

	// 2. Timeout that exceeds
	fmt.Println("2. Timeout that exceeds:")
	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()

	done = make(chan string)
	go func() {
		fmt.Println("  Starting long operation...")
		time.Sleep(3 * time.Second)
		done <- "This won't be received"
	}()

	select {
	case result := <-done:
		fmt.Printf("  Result: %s\n", result)
	case <-ctx.Done():
		fmt.Println("  ✗ Timed out after 1 second!")
	}
	fmt.Println()

	// 3. Context propagation
	fmt.Println("3. Context propagation:")
	ctx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	err := performTask(ctx)
	if err != nil {
		fmt.Printf("  ✗ Error: %v\n", err)
	} else {
		fmt.Println("  ✓ Task completed")
	}
	fmt.Println()

	// 4. Multiple operations with shared timeout
	fmt.Println("4. Multiple operations with shared timeout:")
	ctx, cancel = context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	results := make(chan string, 3)

	// Start 3 operations
	for i := 1; i <= 3; i++ {
		go func(id int) {
			select {
			case <-ctx.Done():
				results <- fmt.Sprintf("Op %d: cancelled", id)
				return
			case <-time.After(time.Duration(id) * time.Second):
				results <- fmt.Sprintf("Op %d: completed", id)
			}
		}(i)
	}

	// Collect results
	for i := 0; i < 3; i++ {
		fmt.Println(" ", <-results)
	}

	fmt.Println()
	fmt.Println("✅ Go timeout complete")
}

func performTask(ctx context.Context) error {
	// Simulate work with context awareness
	select {
	case <-time.After(500 * time.Millisecond):
		fmt.Println("  Task work completed")
		return nil
	case <-ctx.Done():
		return ctx.Err() // context.DeadlineExceeded
	}
}

// Key points:
// - context.WithTimeout for timeout
// - select with ctx.Done() for cancellation
// - Context propagates through function calls
// - cancel() to cleanup
// - More powerful than Ruby Timeout
// - Built-in cancellation support

File: week_25/06_timeout.rb
#!/usr/bin/env ruby
require 'timeout'

# 06. Timeout Pattern in Ruby

puts "=== Ruby Timeout ==="
puts

# 1. Basic timeout
puts "1. Basic timeout:"
begin
  Timeout.timeout(2) do
    puts "  Starting long operation..."
    sleep 1
    puts "  ✓ Completed in time"
  end
rescue Timeout::Error
  puts "  ✗ Operation timed out!"
end
puts

# 2. Timeout that exceeds
puts "2. Timeout that exceeds:"
begin
  Timeout.timeout(1) do
    puts "  Starting long operation..."
    sleep 3
    puts "  This won't print"
  end
rescue Timeout::Error => e
  puts "  ✗ Timed out after 1 second!"
end
puts

# 3. Thread with timeout
puts "3. Thread with timeout:"
thread = Thread.new do
  sleep 5
  "result"
end

# Wait with timeout
result = nil
timeout_thread = Thread.new do
  sleep 1  # 1 second timeout
end

if thread.join(1)  # Wait max 1 second
  result = thread.value
  puts "  ✓ Thread completed: #{result}"
else
  thread.kill
  puts "  ✗ Thread timed out and killed"
end
puts

# 4. Custom timeout with Thread
puts "4. Custom timeout implementation:"

def with_timeout(seconds)
  result = nil
  error = nil
  
  thread = Thread.new do
    begin
      result = yield
    rescue => e
      error = e
    end
  end
  
  if thread.join(seconds)
    raise error if error
    result
  else
    thread.kill
    raise Timeout::Error, "Operation timed out after #{seconds}s"
  end
end

# Success case
begin
  result = with_timeout(2) do
    sleep 0.5
    "Success!"
  end
  puts "  ✓ Result: #{result}"
rescue Timeout::Error => e
  puts "  ✗ #{e.message}"
end

# Timeout case
begin
  result = with_timeout(1) do
    sleep 3
    "Won't complete"
  end
  puts "  Result: #{result}"
rescue Timeout::Error => e
  puts "  ✗ #{e.message}"
end

puts
puts "✅ Ruby timeout complete"

# Key points:
# - Timeout.timeout(seconds) { block }
# - Raises Timeout::Error
# - thread.join(timeout) for manual control
# - thread.kill to force stop (dangerous!)
# - Use for HTTP, DB, long operations
